pinot-core/src/main/java/com/linkedin/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java

```diff
@@ -16,143 +16,602 @@
+                // Something wrong with the controller. Back off and try again.
+                segmentLogger.error("Invalid catchup offset {} in controller response, current offset {}", rspOffset,
+                    _currentOffset);
+                hold();
+              } else {
+                _state = State.CATCHING_UP;
+                _finalOffset = rspOffset;
+                // We will restart consumption when we loop back above.
+              }
+              break;
+            case HOLD:
+              hold();
+              break;
+            case DISCARD:
+              // Keep this in memory, but wait for the online transition, and download when it comes in.
+              _state = State.DISCARDED;
+              break;
+            case KEEP:
+              _state = State.RETAINING;
+              success = buildSegmentAndReplace();
+              if (success) {
+                _state = State.RETAINED;
+              } else {
+                // Could not build segment for some reason. We can only download it.
+                _state = State.ERROR;
+              }
+              break;
+            case COMMIT:
+              _state = State.COMMITTING;
+              success = buildSegment(true);
+              if (!success) {
+                // We could not build the segment. Go into error state.
+                _state = State.ERROR;
+              } else {
+                success = commitSegment();
+                if (success) {
+                  _state = State.COMMITTED;
+                } else {
+                  // Back off and retry from holding state again.
+                  hold();
+                }
+              }
+              break;
+            default:
+              segmentLogger.error("Holding after response from Controller: {}", response.toJsonString());
+              hold();
+              break;
+          }
+        }
+      } catch (Exception e) {
+        segmentLogger.error("Exception while in work", e);
+        _state = State.ERROR;
+      }
+    }
+  }
+
+  private File makeSegmentDirPath() {
+    return new File(_resourceDataDir, _segmentZKMetadata.getSegmentName());
+  }
+
+  private boolean buildSegment(boolean buildTgz) {
+    // Build a segment from in-memory rows.If buildTgz is true, then build the tar.gz file as well
+    File tempSegmentFolder = new File(_resourceTmpDir, "tmp-" + _segmentNameStr + "-" + String.valueOf(now()));
+
+    // lets convert the segment now
+    RealtimeSegmentConverter converter =
+        new RealtimeSegmentConverter(_realtimeSegment, tempSegmentFolder.getAbsolutePath(), _schema,
+            _segmentZKMetadata.getTableName(), _segmentZKMetadata.getSegmentName(), _sortedColumn, _invertedIndexColumns);
+
+    logStatistics();
+    segmentLogger.info("Trying to build segment");
+    final long buildStartTime = now();
+    try {
+      converter.build(_segmentVersion);
+    } catch (Exception e) {
+      segmentLogger.error("Could not build segment", e);
```

Comment:
Good catch. 

I wanted to get exceptions from each step, so i didnt add a finally.